File: frontend/src/types/property.ts
export type ScoreTier = 'hot' | 'warm' | 'cold'

export interface Score {
  total_score: number
  score_tier: ScoreTier | null
  score_breakdown: Record<string, number> | null
  indicator_count: number
  last_scored_at: string
}

export interface Indicator {
  id: number
  indicator_type: string
  status: 'active' | 'resolved' | 'expired'
  source: string
  source_url: string | null
  amount_cents: number | null
  filing_date: string | null
  expiry_date: string | null
  case_number: string | null
  created_at: string
}

export interface Owner {
  name_raw: string
  mailing_address: string | null
  mailing_city: string | null
  mailing_state: string | null
  mailing_zip: string | null
  owner_type: string | null
  is_absentee: boolean | null
  is_out_of_state: boolean | null
}

export interface Listing {
  source: string
  list_price: number | null
  original_price: number | null
  days_on_market: number | null
  price_reductions: number
  listing_status: string | null
  listed_date: string | null
  last_price_cut: string | null
}

export interface County {
  fips_code: string
  name: string
  state_abbr: string
}

export interface PropertyListItem {
  id: number
  address_raw: string
  address_line1: string | null
  address_city: string | null
  address_state: string | null
  address_zip: string | null
  property_type: string | null
  assessed_value: number | null
  score: Score | null
  active_indicator_types: string[]
  lat: number | null
  lng: number | null
}

export interface PropertyDetail extends PropertyListItem {
  apn: string | null
  year_built: number | null
  sqft: number | null
  bedrooms: number | null
  bathrooms: number | null
  lot_size_sqft: number | null
  zoning: string | null
  market_value: number | null
  last_sale_date: string | null
  last_sale_price: number | null
  data_source: string | null
  created_at: string
  updated_at: string
  county: County | null
  owners: Owner[]
  indicators: Indicator[]
  listings: Listing[]
}

export interface PropertyListResponse {
  total: number
  page: number
  page_size: number
  results: PropertyListItem[]
}

export interface MapPoint {
  id: number
  lat: number
  lng: number
  tier: ScoreTier
  score: number
}

export interface GeoJsonFeatureCollection {
  type: 'FeatureCollection'
  features: Array<{
    type: 'Feature'
    geometry: { type: 'Point'; coordinates: [number, number] }
    properties: { id: number; tier: ScoreTier; score: number }
  }>
}

File: backend/app/schemas/property.py
"""Pydantic response schemas for property API endpoints."""

from __future__ import annotations

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class OwnerSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    name_raw: str
    mailing_address: str | None
    mailing_city: str | None
    mailing_state: str | None
    mailing_zip: str | None
    owner_type: str | None
    is_absentee: bool | None
    is_out_of_state: bool | None


class IndicatorSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    indicator_type: str
    status: str
    source: str
    source_url: str | None
    amount_cents: int | None
    filing_date: date | None
    expiry_date: date | None
    case_number: str | None
    created_at: datetime

    @property
    def amount_dollars(self) -> float | None:
        return self.amount_cents / 100 if self.amount_cents else None


class ScoreSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    total_score: float
    score_tier: str | None
    score_breakdown: dict | None
    indicator_count: int
    last_scored_at: datetime


class ListingSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    source: str
    list_price: int | None
    original_price: int | None
    days_on_market: int | None
    price_reductions: int
    listing_status: str | None
    listed_date: date | None
    last_price_cut: date | None


class CountySchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    fips_code: str
    name: str
    state_abbr: str


class PropertyListItem(BaseModel):
    """Lightweight property representation for list view."""
    model_config = ConfigDict(from_attributes=True)

    id: int
    address_raw: str
    address_line1: str | None
    address_city: str | None
    address_state: str | None
    address_zip: str | None
    property_type: str | None
    assessed_value: int | None
    score: ScoreSchema | None
    active_indicator_types: list[str] = []
    lat: float | None = None
    lng: float | None = None


class PropertyDetail(BaseModel):
    """Full property detail for the slide-in panel."""
    model_config = ConfigDict(from_attributes=True)

    id: int
    apn: str | None
    address_raw: str
    address_line1: str | None
    address_city: str | None
    address_state: str | None
    address_zip: str | None
    property_type: str | None
    year_built: int | None
    sqft: int | None
    bedrooms: int | None
    bathrooms: float | None
    lot_size_sqft: int | None
    zoning: str | None
    assessed_value: int | None
    market_value: int | None
    last_sale_date: date | None
    last_sale_price: int | None
    data_source: str | None
    created_at: datetime
    updated_at: datetime

    county: CountySchema | None
    owners: list[OwnerSchema]
    indicators: list[IndicatorSchema]
    score: ScoreSchema | None
    listings: list[ListingSchema]

    lat: float | None = None
    lng: float | None = None


class PropertyListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    results: list[PropertyListItem]


class FilterParams(BaseModel):
    """Query parameters for property list endpoint."""
    state: str | None = None
    # Comma-separated FIPS codes, e.g. "12057" or "12057,12105"
    county_fips: str | None = None
    # Comma-separated zip codes
    zip_codes: str | None = None
    # Comma-separated indicator types
    indicator_types: str | None = None
    score_min: float | None = None
    score_tier: str | None = None
    property_type: str | None = None
    year_built_min: int | None = None
    year_built_max: int | None = None
    assessed_min: int | None = None
    assessed_max: int | None = None
    bedrooms_min: int | None = None
    dom_min: int | None = None
    out_of_state_only: bool = False
    lat: float | None = None
    lng: float | None = None
    radius_km: float | None = None
    sort_by: str = "score_desc"
    page: int = 1
    page_size: int = 50

    def to_filter_dict(self) -> dict:
        d = self.model_dump(exclude_none=True, exclude={"page", "page_size"})
        # Split comma-separated list params into actual lists
        for key in ("county_fips", "zip_codes", "indicator_types"):
            if key in d and isinstance(d[key], str):
                d[key] = [v.strip() for v in d[key].split(",") if v.strip()]
                if not d[key]:
                    del d[key]
        return d
